README.rst
After creating a copy of this template, change the name of the package in `setup.py`, `setup.cfg`, `pytest.ini`, `MANIFEST.in`, `codebuild_formation.yaml` and the subfolder within the `src` directory.  In the Readme, change the all the `python-github-template` in links to point to the new repo name.

Steps to create repo:
   - In Actions -> Dev: click Run workflow. Wait until workflow finishes
   - In Settings -> Security & analysis: enable Dependabot security updates
   - In Setting -> Options, under Merge Button:
      - Make sure "automatically delete head branches" is checked
      - Make sure squash merging" is NOT checked
   - In Settings -> Branches:
      - Add Rule with [dm][ea][vs][et][le][or]* specified as Branch pattern name
         - check Require pull requests reviews before merging
         - check Dismiss stale pull requests
         - check Require Review from Code Owners
         - check Require status checks before merging
         - Under status checks, check all of the python checks (6 total)
         - check Include administrators
         - check Restrict who can push to matching branches

.. image:: https://img.shields.io/pypi/v/python-github-template.svg
    :target: https://pypi.org/project/python-github-template/

.. image:: https://pepy.tech/badge/python-github-template
  :target: https://pepy.tech/project/python-github-template

.. image:: https://img.shields.io/pypi/pyversions/python-github-template.svg
    :target: https://pypi.org/project/python-github-template/

.. image:: https://github.com/CuriBio/python-github-template/workflows/Dev/badge.svg?branch=development
   :alt: Development Branch Build

.. image:: https://codecov.io/gh/CuriBio/python-github-template/branch/development/graph/badge.svg
  :target: https://codecov.io/gh/CuriBio/python-github-template

..
   If this library uses readthedocs then put that badge here
   .. image:: https://readthedocs.org/projects/python-github-template/badge/?version=latest
     :target: https://python-github-template.readthedocs.io/en/latest/?badge=latest
     :alt: Documentation Status


.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black

.. image:: https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white
   :target: https://github.com/pre-commit/pre-commit
   :alt: pre-commit

CHANGELOG.rst
Changelog for Mantarray Waveform Analysis
=========================================


1.3.0 (2021-11-22)
------------------

- Added ability to specify Beta 1/2 mode in ``PipelineTemplate`` and
  and added ability to convert Beta 2 raw memsic data to displacement
- Updated all float32 outputs to float64


1.2.0 (2021-11-05)
------------------

- Updated all analysis to require units of time in microseconds.


1.1.1 (2021-09-30)
------------------

- Fixed new metrics from previous release.


1.1.0 (2021-08-31)
------------------

- Added the following metrics:

  - Ratio of amplitude of each peak to the that of the largest peak.
  - Time from X% width to peak.
  - Time from peak to X% width.


1.0.3 (2021-07-21)
------------------

- Added ability to invert peak detection for magnetic data.


1.0.1 (2021-07-15)
------------------

- Added ``metrics_to_create`` kwarg to ``data_metrics`` to specify which metrics to create. This kwarg can
  also be passed to Pipeline methods that get data metrics.


1.0.0 (2021-05-06)
------------------

- Removed ``filtered_data`` serialized data from ``PeakDectectionError`` subclassed exceptions and from ``find_twitch_indices`` function arguments for performance reasons.


0.9.0 (2021-04-26)
------------------

- Added Twitch Interval Irregularity metric to per twitch metrics sheet and aggregate metrics sheet.


0.8.1 (2021-04-23)
------------------

- Fixed the out of bounds error present in h5 file due to the fix in the  ``peak_detector`` which mitigated the TwoValleysInARow error.


0.8.0 (2021-04-21)
------------------

- Fixed the directionality of the twitch depending on whether the data metrics in the pipeline template are magnetic or force metrics.


0.7.1 (2021-04-01)
------------------

- Fixed TwoValleysInARowError present in h5 file due to ``peak_detector`` in which the second peak is taller than the second.


0.7.0 (2021-03-17)
------------------

- Updated calculations to convert magnetic data to Force


0.6.0 (2021-03-01)
------------------

- Added twitch contraction and relaxation velocity metrics to metric dicitonaries


0.5.11 (2021-02-17)
------------------

- Fixed TwoValleysInARowError present in h5 file due to ``peak_detector`` function


0.5.10 (2020-11-17)
------------------

- Updated dependencies to be compatible with Python 3.9


0.5.8 (2020-11-10)
------------------

- Fixed issue with peak detection on data with no detected valleys.


0.5.6 (2020-11-04)
------------------

- Fixed ``publish`` job.
- Publishing using github workflow and build environment.
- Fixed incorrect raising of TwoValleysInARowError.
- Fixed issue with two valleys incorrectly being found between peaks.


0.5.4 (2020-09-30)
------------------

- Fixed peak detection so it is less likely to report two peaks/valleys in a row.


0.5.3 (2020-09-15)
------------------

- Added TwoValleysInARowError.
- Fixed TwoPeaksInARowError reporting.


0.5.2 (2020-09-09)
------------------

- Added upload of source files to pypi for linux python3.7 download.


0.5.1 (2020-09-09)
------------------

- Added 30 Hz Butterworth Filter.


0.5.0 (2020-09-08)
------------------

- Added Twitch Frequency metric.
- Added peak detetection and metric calculation (for magnetic signal) to Pipeline.
- Created alias of load_raw_magnetic_data to become more agnostic to sensor type.


0.4.1 (2020-09-02)
------------------

- Added 30 Hz Low-Pass Bessel filter.
- Added small speed upgrade to cython compression code.


0.4.0 (2020-09-01)
------------------

- Refactored twitch width analysis so that it interpolates to find a point to use.
- Added aggregate statistic metrics for twitch widths.
- Refactored peak detection to be more robust.
- Cached the filter coefficients in PipelineTemplate to improve performance.


0.3.1 (2020-08-31)
------------------

- Added compression speed improvements.
- Fixed edge case in compression for horizontal line r squared.


src/mantarray_waveform_analysis/__init__.py
# -*- coding: utf-8 -*-
"""Docstring."""
from . import pipelines
from . import transforms
from .constants import ADC_GAIN
from .constants import ALL_METRICS
from .constants import AMPLITUDE_UUID
from .constants import AUC_UUID
from .constants import BESSEL_BANDPASS_UUID
from .constants import BESSEL_LOWPASS_10_UUID
from .constants import BESSEL_LOWPASS_30_UUID
from .constants import BUTTERWORTH_LOWPASS_30_UUID
from .constants import CENTIMILLISECONDS_PER_SECOND
from .constants import CONTRACTION_TIME_UUID
from .constants import CONTRACTION_VELOCITY_UUID
from .constants import FRACTION_MAX_UUID
from .constants import GAUSS_PER_MILLITESLA
from .constants import IRREGULARITY_INTERVAL_UUID
from .constants import MEMSIC_CENTER_OFFSET
from .constants import MEMSIC_FULL_SCALE
from .constants import MEMSIC_MSB
from .constants import MICRO_TO_BASE_CONVERSION
from .constants import MIDSCALE_CODE
from .constants import MILLI_TO_BASE_CONVERSION
from .constants import MILLIMETERS_PER_MILLITESLA
from .constants import MILLIVOLTS_PER_MILLITESLA
from .constants import MIN_NUMBER_PEAKS
from .constants import MIN_NUMBER_VALLEYS
from .constants import NEWTONS_PER_MILLIMETER
from .constants import PRIOR_PEAK_INDEX_UUID
from .constants import PRIOR_VALLEY_INDEX_UUID
from .constants import RAW_TO_SIGNED_CONVERSION_VALUE
from .constants import REFERENCE_VOLTAGE
from .constants import RELAXATION_TIME_UUID
from .constants import RELAXATION_VELOCITY_UUID
from .constants import SUBSEQUENT_PEAK_INDEX_UUID
from .constants import SUBSEQUENT_VALLEY_INDEX_UUID
from .constants import TIME_DIFFERENCE_UUID
from .constants import TWITCH_FREQUENCY_UUID
from .constants import TWITCH_PERIOD_UUID
from .constants import WIDTH_FALLING_COORDS_UUID
from .constants import WIDTH_RISING_COORDS_UUID
from .constants import WIDTH_UUID
from .constants import WIDTH_VALUE_UUID
from .exceptions import DataAlreadyLoadedInPipelineError
from .exceptions import FilterCreationNotImplementedError
from .exceptions import PeakDetectionError
from .exceptions import TooFewPeaksDetectedError
from .exceptions import TwoPeaksInARowError
from .exceptions import TwoValleysInARowError
from .exceptions import UnrecognizedFilterUuidError
from .exceptions import UnsupportedTransformationError
from .peak_detection import find_twitch_indices
from .peak_detection import peak_detector
from .pipelines import Pipeline
from .pipelines import PipelineTemplate
from .transforms import apply_empty_plate_calibration
from .transforms import apply_noise_filtering
from .transforms import apply_sensitivity_calibration
from .transforms import calculate_displacement_from_magnetic_flux_density
from .transforms import calculate_displacement_from_voltage
from .transforms import calculate_force_from_displacement
from .transforms import calculate_magnetic_flux_density_from_memsic
from .transforms import calculate_voltage_from_gmr
from .transforms import create_filter
from .transforms import FILTER_CHARACTERISTICS
from .transforms import noise_cancellation

# Kristian 9/15/21


if 5 < 10:  # pragma: no cover # protect this from zimports deleting the pylint disable statement
    from .compression_cy import (  # pylint: disable=import-error # Eli (8/18/20) unsure why pylint is unable to recognize cython import... # Tanner (8/31/20) Pylint also flags this as duplicate code due to a similar import in pipelines.py, which may be related to pylint's import issues
        compress_filtered_magnetic_data,
    )

__all__ = [
    "transforms",
    "pipelines",
    "TWITCH_PERIOD_UUID",
    "TWITCH_FREQUENCY_UUID",
    "PRIOR_PEAK_INDEX_UUID",
    "PRIOR_VALLEY_INDEX_UUID",
    "SUBSEQUENT_PEAK_INDEX_UUID",
    "SUBSEQUENT_VALLEY_INDEX_UUID",
    "AMPLITUDE_UUID",
    "AUC_UUID",
    "WIDTH_UUID",
    "WIDTH_VALUE_UUID",
    "WIDTH_RISING_COORDS_UUID",
    "WIDTH_FALLING_COORDS_UUID",
    "CENTIMILLISECONDS_PER_SECOND",
    "MIDSCALE_CODE",
    "RAW_TO_SIGNED_CONVERSION_VALUE",
    "apply_sensitivity_calibration",
    "noise_cancellation",
    "apply_empty_plate_calibration",
    "apply_noise_filtering",
    "create_filter",
    "UnrecognizedFilterUuidError",
    "FilterCreationNotImplementedError",
    "DataAlreadyLoadedInPipelineError",
    "TwoPeaksInARowError",
    "TwoValleysInARowError",
    "BESSEL_BANDPASS_UUID",
    "BESSEL_LOWPASS_10_UUID",
    "FILTER_CHARACTERISTICS",
    "compress_filtered_magnetic_data",
    "calculate_voltage_from_gmr",
    "calculate_displacement_from_voltage",
    "PipelineTemplate",
    "Pipeline",
    "peak_detector",
    "find_twitch_indices",
    "TooFewPeaksDetectedError",
    "MIN_NUMBER_PEAKS",
    "MIN_NUMBER_VALLEYS",
    "BESSEL_LOWPASS_30_UUID",
    "BUTTERWORTH_LOWPASS_30_UUID",
    "RELAXATION_VELOCITY_UUID",
    "CONTRACTION_VELOCITY_UUID",
    "calculate_force_from_displacement",
    "MILLI_TO_BASE_CONVERSION",
    "MILLIVOLTS_PER_MILLITESLA",
    "MILLIMETERS_PER_MILLITESLA",
    "NEWTONS_PER_MILLIMETER",
    "REFERENCE_VOLTAGE",
    "ADC_GAIN",
    "IRREGULARITY_INTERVAL_UUID",
    "ALL_METRICS",
    "FRACTION_MAX_UUID",
    "TIME_DIFFERENCE_UUID",
    "RELAXATION_TIME_UUID",
    "CONTRACTION_TIME_UUID",
    "MICRO_TO_BASE_CONVERSION",
    "PeakDetectionError",
    "MEMSIC_CENTER_OFFSET",
    "MEMSIC_MSB",
    "MEMSIC_FULL_SCALE",
    "GAUSS_PER_MILLITESLA",
    "calculate_magnetic_flux_density_from_memsic",
    "calculate_displacement_from_magnetic_flux_density",
    "UnsupportedTransformationError",
]
